sdk/spring/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/core/CosmosTemplate.java

```diff
@@ -856,13 +858,40 @@ public <T, ID> Iterable<T> findByIds(Iterable<ID> ids, Class<T> domainType, Stri
         Assert.notNull(ids, "Id list should not be null");
         Assert.notNull(domainType, "domainType should not be null.");
         Assert.hasText(containerName, "container should not be null, empty or only whitespaces");
-        final List<Object> idList = new ArrayList<>();
-        for (ID id : ids) {
-            idList.add(CosmosUtils.getStringIDValue(id));
+
+        CosmosEntityInformation<?, ?> cosmosEntityInformation = CosmosEntityInformation.getInstance(domainType);
+        String containerPartitionKey = cosmosEntityInformation.getPartitionKeyFieldName();
+        if ("id".equals(containerPartitionKey) && ids.iterator().next() != null) {
```

Comment:
What exception is thrown when partitionKey is not id? Looks like either NPE or "Failed to find items"? Error should make it obvious that this method only supports entities where pk is id (and a unit test checking for this?).